substrate-archive/src/database.rs

```diff
@@ -269,6 +271,117 @@ impl Insert for Metadata {
+
+		for span in self.spans.into_iter() {
+			let id = i32::try_from(span.id.into_u64())?;
+			let parent_id: Option<i32> =
+				if let Some(id) = span.parent_id { Some(i32::try_from(id.into_u64())?) } else { None };
+			let overall_time = shave_nanos(time_to_std(span.overall_time)?)?;
+			batch.reserve(11)?;
+			if batch.current_num_arguments() > 0 {
+				batch.append(",");
+			}
+			batch.append("(");
+			batch.bind(block_num)?; // block_numk
+			batch.append(",");
+			batch.bind(hash.as_slice())?; // hash
+			batch.append(",");
+			batch.bind(false)?; // is_event
+			batch.append(",");
+			batch.bind(span.start_time)?;
+			batch.append(",");
+			batch.bind(overall_time)?; // timestamp
+			batch.append(",");
+			batch.bind(span.file)?; // file
+			batch.append(",");
+			batch.bind(span.line)?; // line
+			batch.append(",");
+			batch.bind(id)?; // trace_id
+			batch.append(",");
+			batch.bind(parent_id)?; // trace_parent_id
+			batch.append(",");
+			batch.bind(span.target)?; // target
+			batch.append(",");
+			batch.bind(span.name)?; // name
+			batch.append(",");
+			batch.bind(sqlx::types::Json(span.values))?; // traces
+			batch.append(")");
+		}
+
+		for event in self.events.into_iter() {
+			let parent_id = Some(i32::try_from(event.parent_id.into_u64())?);
+			batch.reserve(11)?;
+			if batch.current_num_arguments() > 0 {
+				batch.append(",");
+			}
+			batch.append("(");
+			batch.bind(block_num)?; // block number
+			batch.append(",");
+			batch.bind(hash.as_slice())?; // hash
+			batch.append(",");
+			batch.bind(true)?; // is_event
+			batch.append(",");
+			batch.bind(event.time)?; //time
+			batch.append(",");
+			batch.bind(Option::<chrono::Duration>::None)?; // an event won't have a duration
+			batch.append(",");
+			batch.bind(event.file)?; // file
+			batch.append(",");
+			batch.bind(event.line)?; // line
+			batch.append(",");
+			batch.bind(Option::<i32>::None)?; // Event has no ID
+			batch.append(",");
+			batch.bind(parent_id)?; // parent ikd
+			batch.append(",");
+			batch.bind(event.target)?; // target
+			batch.append(",");
+			batch.bind(event.name)?; // name
+			batch.append(",");
+			batch.bind(sqlx::types::Json(event.values))?; // values
+			batch.append(")");
+		}
+
+		Ok(batch.execute(conn).await?)
+	}
+}
+
+// Chrono depends on an error type in `time` that is a full version behind the one that SQLX uses
+// This function avoids depending on two time lib.
+// Old time is disabled in chrono by not providing the feature flag in Cargo.toml.
+fn time_to_std(time: chrono::Duration) -> Result<std::time::Duration> {
+	if time < chrono::Duration::zero() {
+		Err(ArchiveError::TimestampOutOfRange)
+	} else {
+		Ok(time.to_std().expect("Checked for less than 0"))
+	}
+}
+
+// the alternative to this is creating a Composite postgres type
+// that stores seconds + extra nanoseconds.
+// Postgres does not support nanosecond precision.
+fn shave_nanos(time: std::time::Duration) -> Result<std::time::Duration> {
```

Comment:
what about casting `Duration::as_nanos()` from `u128` to `i64` ? then could be stored in a `bigint`? Still allows for 292 years duration at nanosecond precision.

Reply:
Could definitely do that -- I didn't do it the first time around because I liked that it was in an 'official' Postgres time format. However if it's just nanos that's fairly simple to convert back and forth

Reply:
yeah, if nanos aren't important then using interval def has some advantages - there's also `Duration::as_secs_f64` which would avoid a cast

Reply:
There's some tension here between making the data easy to use (native PG time format is good here) and high precision. As the primary use case for the time being is balance reconciliation I'd say we stick with the native time format for now and perhaps follow up by adding a new column with the full-precision timestamp (or add it right now if it's easy to do)?